## Course Assignment Instructions
You should have Python (version 3.8 or later) and Jupyter Notebook installed to complete this assignment. You will write code in the empty cell/cells below the problem. While most of this will be a programming assignment, some questions will ask you to "write a few sentences" in markdown cells. 

Submission Instructions:

Create a labs directory in your personal class repository (e.g., located in your home directory)
Clone the class repository
Copy this Jupyter notebook file (.ipynb) into your repo/labs directory
Make your edits, commit changes, and push to your repository
All submissions must be pushed before the due date to avoid late penalties. 

Labs are graded out of a 100 pts. Each day late is -5. For a max penalty of -50 after 10 days. From there you may submit the lab anytime before the semester ends for a max score of 50.  

Lab 2 is due on 2/18/25

## Basic Modeling
In the 342 class an example was given that considered a variable `x_3` which measured "criminality". In this example there are L = 4 levels "none", "infraction", "misdemeanor" and "felony". Create a variable `x_3` here with 100 random elements (equally probable). Create it as a nominal (i.e. unordered) factor. Hint: use random.choice from NumPy and Categorical from Pandas.

In [12]:
import numpy as np
import pandas as pd

# Define the categories

categories = ["none","infraction","misdemeanor","felony"]


# Generate 100 random elements with equal probability

x_3 = np.random.choice(categories, size = 100, replace = True)

# Convert to a categorical (nominal- "nominal" refers to a type of 
#categorical data where values represent distinct categories 
#without any inherent order or ranking) variable in pandas

x_3 = pd.Categorical(x_3, categories = categories, ordered = False)
x_3



['misdemeanor', 'misdemeanor', 'felony', 'misdemeanor', 'misdemeanor', ..., 'none', 'none', 'misdemeanor', 'none', 'misdemeanor']
Length: 100
Categories (4, object): ['none', 'infraction', 'misdemeanor', 'felony']

Use x_3 to create x_3_bin, a binary feature where 0 is no crime and 1 is any crime.

In [13]:
# creates a boolean array (True for crime, False for no crime)
x_3_bin = (x_3!="none").astype(int)
x_3_bin


array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1])

Use `x_3` to create `x_3_ord`, an ordered factor variable. Ensure the proper ordinal ordering.

In [14]:
categories = ["infraction","none","misdemeanor","felony"]
x_3_ord=pd.Categorical(x_3, categories = categories, ordered = True)
x_3_ord

['misdemeanor', 'misdemeanor', 'felony', 'misdemeanor', 'misdemeanor', ..., 'none', 'none', 'misdemeanor', 'none', 'misdemeanor']
Length: 100
Categories (4, object): ['infraction' < 'none' < 'misdemeanor' < 'felony']

Convert this variable into three binary variables without any information loss and put them into a data matrix. Hint: use column_stack from Numpy.

In [15]:
# takes all 3 arrays and makes matrix of it/stacks it together
x_3_matrix = np.column_stack([
    (x_3 == "infraction").astype(int),
    (x_3 == "felony").astype(int),
    (x_3 == "misdemeanor").astype(int)
]
)
# now convert to matrix using pandas

x_3_matrix = pd.DataFrame(x_3_matrix, columns = ["infraction","misdemeanor","felony"])
x_3_matrix

infraction  misdemeanor  felony
0            0            0       1
1            0            0       1
2            0            1       0
3            0            0       1
4            0            0       1
..         ...          ...     ...
95           0            0       0
96           0            0       0
97           0            0       1
98           0            0       0
99           0            0       1

[100 rows x 3 columns]

What should the sum of each row be (in English)? Write your answer in the markdown cell below

 each row should be 1 or 0 so sum will be 1 

Verify that in the code cell below

In [22]:
row_sum = x_3_matrix.sum(axis = 1)
#axis = 1 is row and =0 is columns
row_sum
print(row_sum.value_counts())

1    82
0    18
Name: count, dtype: int64


 How should the column sum look (in English)? Write your answer in the markdown cell below

atleast 0 atmost n - approximately 25 per column cb we sampled 4 categories evenly from 100 samples

Verify that in the code cell below

In [27]:
col_sum = x_3_matrix.sum(axis = 0) 
print(col_sum)

infraction     27
misdemeanor    27
felony         28
dtype: int64


Generate a matrix with 100 rows where the first column is realization from a normal with mean 17 and variance 38, the second column is uniform between -10 and 10, the third column is poisson with mean 6, the fourth column is exponential with lambda of 9, the fifth column is binomial with n = 20 and p = 0.12 and the sixth column is a binary variable with exactly 24% 1's dispersed randomly. Name the rows the entries of the `fake_first_names` vector. You will need to use Numpy

In [31]:
# Number of rows
num_rows = 100

# Assign row names (index) from fake_first_names
fake_first_names = [
    "Sophia", "Emma", "Olivia", "Ava", "Mia", "Isabella", "Riley", 
    "Aria", "Zoe", "Charlotte", "Lily", "Layla", "Amelia", "Emily", 
    "Madelyn", "Aubrey", "Adalyn", "Madison", "Chloe", "Harper", 
    "Abigail", "Aaliyah", "Avery", "Evelyn", "Kaylee", "Ella", "Ellie", 
    "Scarlett", "Arianna", "Hailey", "Nora", "Addison", "Brooklyn", 
    "Hannah", "Mila", "Leah", "Elizabeth", "Sarah", "Eliana", "Mackenzie", 
    "Peyton", "Maria", "Grace", "Adeline", "Elena", "Anna", "Victoria", 
    "Camilla", "Lillian", "Natalie", "Jackson", "Aiden", "Lucas", 
    "Liam", "Noah", "Ethan", "Mason", "Caden", "Oliver", "Elijah", 
    "Grayson", "Jacob", "Michael", "Benjamin", "Carter", "James", 
    "Jayden", "Logan", "Alexander", "Caleb", "Ryan", "Luke", "Daniel", 
    "Jack", "William", "Owen", "Gabriel", "Matthew", "Connor", "Jayce", 
    "Isaac", "Sebastian", "Henry", "Muhammad", "Cameron", "Wyatt", 
    "Dylan", "Nathan", "Nicholas", "Julian", "Eli", "Levi", "Isaiah", 
    "Landon", "David", "Christian", "Andrew", "Brayden", "John", 
    "Lincoln"
]

# Create a DataFrame with the specified distributions

X = pd.DataFrame(
    {
    "Normal": np.random.normal(loc=17, scale = np.sqrt(38), size=num_rows),  # Normal(17, variance 38)
    "Uniform": np.random.uniform(low= -1, high=10, size=num_rows),                 # Uniform(-10, 10)
    "Poisson": np.random.poisson(6,size=num_rows),                            # Poisson(6)
    "Exponential": np.random.exponential((1/9),size=num_rows),                  # Exponential(λ=9)
    "Binomial": np.random.binomial(n=20, p=0.12,size=num_rows),             # Binomial(n=20, p=0.12)
    "Binary": np.random.permutation( [1]*int(num_rows*0.24) + [0]*int(num_rows*0.76))  # 24% 1s, shuffled
}
)

X.index = fake_first_names[:num_rows]
X



Normal   Uniform  Poisson  Exponential  Binomial  Binary
Sophia     19.146136  8.059100        9     0.246700         2       0
Emma       20.248167  0.657792        7     0.040585         1       0
Olivia     17.212600  1.664471        8     0.079142         1       0
Ava        23.937950 -0.799742        3     0.105012         5       0
Mia        10.306851  1.941303        7     0.123583         0       0
...              ...       ...      ...          ...       ...     ...
Christian  18.757201  7.606783        3     0.166500         7       0
Andrew     10.981748 -0.557392        5     0.037886         1       0
Brayden     7.108700  9.718888        7     0.068551         3       0
John       14.021473  0.948379        4     0.107377         2       0
Lincoln    17.295381  7.376450        3     0.204942         1       1

[100 rows x 6 columns]

Create a data frame of the same data as above except make the binary variable a factor "DOMESTIC" vs "FOREIGN" for 0 and 1 respectively. In Rstudio you used the `View` function to ensure this worked as desired. In python use .head() on the DataFrame. I recommend creating a copy of the DataFrame and then using the .replace in conjunction with .astype("category") to make the binary variable a factor. 

In [33]:
# Convert matrix DataFrame to categorical for the binary variable
# Make a copy to keep X unchanged- we wanna switch binary col where we have 1's and 0's
x_copy=X.copy()


# Convert binary column (6th column) to categorical labels
x_copy["Binary"]= x_copy["Binary"].replace({0:"Domestic",1:"Foreign"}).astype("category")

# Display first few rows
x_copy.head()


Normal   Uniform  Poisson  Exponential  Binomial    Binary
Sophia  19.146136  8.059100        9     0.246700         2  Domestic
Emma    20.248167  0.657792        7     0.040585         1  Domestic
Olivia  17.212600  1.664471        8     0.079142         1  Domestic
Ava     23.937950 -0.799742        3     0.105012         5  Domestic
Mia     10.306851  1.941303        7     0.123583         0  Domestic

Print out a table of the binary variable. Then print out the proportions of "DOMESTIC" vs "FOREIGN". Pandas DataFrames has a .value_counts() feature. 

In [37]:
x_copy["Binary"].value_counts(normalize=True)

Binary
Domestic    0.76
Foreign     0.24
Name: proportion, dtype: float64

Print out a summary of the whole dataframe.

In [41]:
print(x_copy.describe())
print(x_copy["Binary"].value_counts())

           Normal     Uniform     Poisson  Exponential    Binomial
count  100.000000  100.000000  100.000000   100.000000  100.000000
mean    17.358323    3.961668    5.870000     0.110200    2.650000
std      5.951614    3.153289    2.480856     0.114190    1.641476
min      1.487877   -0.826182    1.000000     0.000303    0.000000
25%     13.988447    1.228443    4.000000     0.028539    1.000000
50%     17.206678    3.567837    6.000000     0.065005    2.500000
75%     21.039802    7.191022    8.000000     0.150016    3.000000
max     31.762679    9.990499   12.000000     0.658182    7.000000
Binary
Domestic    76
Foreign     24
Name: count, dtype: int64


## Dataframe creation
Imagine you are running an experiment with many manipulations. You have 14 levels in the variable "treatment" with levels a, b, c, etc. For each of those manipulations you have 3 submanipulations in a variable named "variation" with levels A, B, C. Then you have "gender" with levels M / F. Then you have "generation" with levels Boomer, GenX, Millenial. Then you will have 6 runs per each of these groups. In each set of 6 you will need to select a name without duplication from the appropriate set of names (from the last question). Create a data frame with columns treatment, variation, gender, generation, name and y that will store all the unique unit information in this experiment. Leave y empty because it will be measured as the experiment is executed. In Rstudio you used `rep` function using the `times` argument. For python use np.tile, and np.repeat.

In [102]:

# he will post code on discord

# Define categories 
treatments = list("abcdefghijklmn") 

# 14 levels
variations = list("ABC") 

# 3 levels 
genders = ["M", "F"] 

# 2 levels 
generations = ["Boomer", "GenX", "Millenial"] 

# 3 levels 
# Define name sets 
name_sets = { "M": { "Boomer": ["Theodore", "Bernard", "Gene", "Herbert", "Ray", 
"Tom", "Lee", "Alfred", "Leroy", "Eddie"], "GenX": ["Marc", "Jamie", "Greg", "Darryl",
"Tim", "Dean", "Jon", "Chris", "Troy", "Jeff"], "Millenial": ["Zachary", "Dylan", "Christian",
"Wesley", "Seth", "Austin", "Gabriel", "Evan", "Casey", "Luis"] }, "F": { "Boomer": ["Gloria", "Joan", "Dorothy", 
"Shirley", "Betty", "Dianne", "Kay", "Marjorie", "Lorraine", "Mildred"], "GenX": ["Tracy", "Dawn", "Tina", "Tammy", 
"Melinda", "Tamara", "Tracey", "Colleen", "Sherri", "Heidi"], "Millenial": ["Samantha", "Alexis", "Brittany", "Lauren", 
"Taylor", "Bethany", "Latoya", "Candice", "Brittney", "Cheyenne"] } }


# Create experiment DataFrame 
df = pd.DataFrame({ "treatment": np.repeat(treatments, len(variations) * len(genders) * len(generations) * 6), 
                   "variation": np.tile(np.repeat(variations, len(genders) * len(generations) * 6), len(treatments)), 
                   "gender": np.tile(np.repeat(genders, len(generations) * 6), len(treatments) * len(variations)), 
                   "generation": np.tile(np.repeat(generations, 6), len(treatments) * len(variations) * len(genders)), }) 

# Add a unique identifier to preserve the original order
df = df.reset_index().rename(columns={'index': 'orig_index'})

# Define a function that assigns 6 unique names per group and returns a DataFrame with the original index
def assign_names_with_index(group):
    gender_val = group["gender"].iloc[0]  # Extract the group's gender
    generation_val = group["generation"].iloc[0]  # Extract the group's generation

    # Sample 6 unique names from the appropriate set (without replacement)
    names = np.random.choice(name_sets[gender_val][generation_val], 6, replace=False)

    # Return a DataFrame with the original indices and the assigned names
    return pd.DataFrame({
        "orig_index": group["orig_index"],
        "name": names
    })

# Group by the categorical variables and apply the function
names_df = df.groupby(["treatment", "variation", "gender", "generation"], group_keys=False).apply(assign_names_with_index).reset_index(drop=True)

# Merge the assigned names back into the original DataFrame using the unique identifier
df = df.merge(names_df, on="orig_index", how="left")

# Restore the original order and remove the temporary identifier
df = df.sort_values("orig_index").reset_index(drop=True).drop(columns=["orig_index"])

# Add an empty column 'y'
df["y"] = np.nan

# Display DataFrame
df

C:\Users\ilana brunner\AppData\Local\Temp\ipykernel_29500\2333051876.py:50: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  names_df = df.groupby(["treatment", "variation", "gender", "generation"], group_keys=False).apply(assign_names_with_index).reset_index(drop=True)


treatment variation gender generation      name   y
0            a         A      M     Boomer   Bernard NaN
1            a         A      M     Boomer     Leroy NaN
2            a         A      M     Boomer       Tom NaN
3            a         A      M     Boomer       Lee NaN
4            a         A      M     Boomer   Herbert NaN
...        ...       ...    ...        ...       ...  ..
1507         n         C      F  Millenial  Brittney NaN
1508         n         C      F  Millenial    Taylor NaN
1509         n         C      F  Millenial   Candice NaN
1510         n         C      F  Millenial  Samantha NaN
1511         n         C      F  Millenial  Cheyenne NaN

[1512 rows x 6 columns]

Now that you've done it with the np.tile and np.repeat, Try doing this by importing product from the itertools module. This will be analogous to using `expand.grid` function from Rstudio. 

| **R Function** | **Python Equivalent** |
|--------------|-----------------|
| `rep(x, times=n)` | `np.repeat(x, n)` |
| `rep(x, each=n)` | `np.tile(np.repeat(x, n), times)` |
| `rep(x, length.out=n)` | `np.resize(x, n)` |
| `expand.grid()` | `itertools.product()` |

| **R Function** | **Python Equivalent** | **Use Case** |
|--------------|-----------------|-----------|
| `rep(x, times=n)` | `np.repeat(x, n)` | Repeat each element **`n` times** in order |
| `rep(x, each=n)` | `np.tile(x, n)` | Repeat the full sequence **`n` times** |
| `rep(x, length.out=n)` | `np.resize(x, n)` | Repeat `x` but **truncate** or **expand** to length `n` |

**`expand.grid()` → `itertools.product()`** for generating **all combinations**  
**`rep(..., each=n)` → `np.repeat()`** for **repeating values in order**  
**`rep(..., times=n)` → `np.tile()`** for **cycling through values**  
**`Combination of `np.repeat()` and `np.tile()`** replaces **nested `rep()`** in R

In [105]:
# he is gonna post code 

import pandas as pd
import numpy as np
from itertools import product

# Define categorical variables
treatments = ["A", "B"]
variations = ["X", "Y"]
genders = ["Male", "Female"]
generations = ["Gen1", "Gen2"]

# Create a DataFrame using itertools.product
df = pd.DataFrame(product(treatments, variations, genders, generations),
                  columns=["treatment", "variation", "gender", "generation"])

# Assign unique index
df["orig_index"] = df.index  

# Define name sets for each gender and generation
name_sets = {
    "Male": {
        "Gen1": ["John", "Michael", "David", "James", "Robert", "William"],
        "Gen2": ["Liam", "Noah", "Oliver", "Elijah", "Lucas", "Mason"]
    },
    "Female": {
        "Gen1": ["Mary", "Patricia", "Jennifer", "Linda", "Barbara", "Elizabeth"],
        "Gen2": ["Olivia", "Emma", "Ava", "Sophia", "Isabella", "Mia"]
    }
}

# Function to assign names dynamically based on group size
def assign_names_with_index(group):
    gender_val = group["gender"].iloc[0]
    generation_val = group["generation"].iloc[0]
    
    group_size = len(group)
    available_names = name_sets[gender_val][generation_val]
    
    # Ensure we don't sample more names than available
    num_names = min(group_size, len(available_names))
    
    # Sample unique names
    names = np.random.choice(available_names, num_names, replace=False)
    
    return pd.DataFrame({
        "orig_index": group["orig_index"].values[:num_names],  
        "name": names
    })

# Apply the function to assign names
names_df = df.groupby(["treatment", "variation", "gender", "generation"], group_keys=False).apply(assign_names_with_index).reset_index(drop=True)

# Merge assigned names back into original DataFrame
df = df.merge(names_df, on="orig_index", how="left").sort_values("orig_index").reset_index(drop=True)

# Add empty column 'y'
df["y"] = np.nan

# Display final DataFrame
print(df)



   treatment variation  gender generation  orig_index       name   y
0          A         X    Male       Gen1           0    Michael NaN
1          A         X    Male       Gen2           1      Mason NaN
2          A         X  Female       Gen1           2  Elizabeth NaN
3          A         X  Female       Gen2           3        Ava NaN
4          A         Y    Male       Gen1           4       John NaN
5          A         Y    Male       Gen2           5      Lucas NaN
6          A         Y  Female       Gen1           6       Mary NaN
7          A         Y  Female       Gen2           7     Olivia NaN
8          B         X    Male       Gen1           8    William NaN
9          B         X    Male       Gen2           9     Elijah NaN
10         B         X  Female       Gen1          10    Barbara NaN
11         B         X  Female       Gen2          11     Sophia NaN
12         B         Y    Male       Gen1          12      James NaN
13         B         Y    Male    

C:\Users\ilana brunner\AppData\Local\Temp\ipykernel_29500\1365038993.py:52: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  names_df = df.groupby(["treatment", "variation", "gender", "generation"], group_keys=False).apply(assign_names_with_index).reset_index(drop=True)


## Basic Binary Classification Modeling

Load the famous `iris` data frame into the namespace. In Rstudio you used the `skim` function from the package `skimr` to provide a summary of the columns. In python we will use df.describe() and the ProfileReport from the ydata-profiling package. The `iris` data set is not available in base python, but we can get this data from the sklearn package. Write a few descriptive sentences about the distributions using the code below in English.

In [44]:
#install scikit-learn by uncommenting the code below
%pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.


In [46]:
#install ydata-profile by uncommenting the code below
%pip install -U ydata-profiling[notebook]

  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
   ---------------------------------------- 0.0/666.4 kB ? eta -:--:--
   ---------------------------------------- 666.4/666.4 kB 8.6 MB/s eta 0:00:00
  Created wheel for htmlmin: filename=htmlmin-0.1.12-py3-none-any.whl size=27091 sha256=f342c5a2276af9e6d0b8c9c923fe663ebb73bfc7b61b6811c229d335725a1775
  Stored in directory: c:\users\ilana brunner\appdata\local\pip\cache\wheels\5f\d4\d7\4189b07b5902ee9f3ce0dbb14909fbe8037c39d6c63ffd49c9
Successfully built htmlmin
Note: you may need to restart the kernel to use updated packages.


### **Comparing the `iris` Dataset in R vs Python**
| Feature  | **R (`datasets::iris`)**  | **Python (`sklearn.datasets.load_iris()`)**  |
|----------|-------------------------|--------------------------------|
| **Total Rows**  | 150 | 150 |
| **Columns (Features)** | 5 (`Sepal.Length`, `Sepal.Width`, `Petal.Length`, `Petal.Width`, `Species`) | 5 (`sepal length (cm)`, `sepal width (cm)`, `petal length (cm)`, `petal width (cm)`, `species`) |
| **Species Encoding**  | `"setosa"`, `"versicolor"`, `"virginica"` (Categorical Factor) | `0` (setosa), `1` (versicolor), `2` (virginica) (Numerical Encoding) |
| **Data Type for Species** | Factor (Categorical) | Integer (0,1,2) |
| **Data Loading Method** | `data(iris)` (built-in dataset) | `datasets.load_iris()` (from `sklearn`) |

### **Key Differences**
- **Species Encoding:**  
  - **R uses categorical factor labels (`setosa`, `versicolor`, `virginica`).**  
  - **Python (`sklearn`) encodes species numerically as `0`, `1`, and `2`.**
- **Column Names:**  
  - **R:** `Sepal.Length`, `Sepal.Width`, etc.  
  - **Python:** `sepal length (cm)`, `sepal width (cm)`, etc.  

In [61]:
from sklearn import datasets
import ydata_profiling  

# Load the famous Iris dataset
Iris = datasets.load_iris()
#get length and width
df_iris = pd.DataFrame(Iris.data, columns = Iris.feature_names)
# now add species column
df_iris["species"]=Iris.target
df_iris.describe()

profile = ydata_profiling.ProfileReport(df_iris, title = "iris summary", explorative= True)
# Generate the profiling report (Uncomment to generate HTML file)
profile.to_file("iris_report.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

C:\Users\ilana brunner\anaconda3\Lib\site-packages\ydata_profiling\model\pandas\discretize_pandas.py:52: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4
 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 9 9 9 9 9 9 9 9 9 9 9
 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9
 9 9]' has dtype incompatible with int32, please explicitly cast to a compatible dtype first.
  discretized_df.loc[:, column] = self._discretize_column(


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

TO-DO: describe this data

Sepal Length:
min= 4.3
max= 7.9
mean = 5.843333

sepal width:

min=2
max=4.4
mean= 3.057333

petal L:
min=1
max=6.9
mean=3.578

petal W:
min=.1
max=2.5
mean=1.993333

3 distinct species 
2.0% distinct


The outcome / label / response is `Species`. This is what we will be trying to predict. However, we only care about binary classification between "setosa" and "versicolor" for the purposes of this exercise. Thus the first order of business is to drop one class. Let's drop the data for the level "virginica" from the data frame.

In [68]:
# Filter out "virginica" from the dataset
df_iris_binary= df_iris[df_iris["species"] != 2 ].copy() # form of filtering
print(df_iris_binary.head())
print(df_iris_binary["species"].unique()) #shows only 2 unique values 0 or 1 aka got rid of virginica


   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   species  
0        0  
1        0  
2        0  
3        0  
4        0  
[0 1]


Now create a vector `y` that is length the number of remaining rows in the data frame whose entries are 0 if "setosa" and 1 if "versicolor".

In [70]:
# Create binary target vector `y` (0 for setosa, 1 for versicolor)
y = (df_iris_binary["species"]==1).astype(int)
y.unique()


array([0, 1])

Write a function `mode` returning the sample mode of a vector of numeric values. Use np.random.choice from NumPy and import Counter from the collections module.

In [77]:
from collections import Counter

# Define mode function
def mode(v):
    return Counter(v).most_common(1)[0][0]

    # Test with a random sample (equivalent to `sample(letters, 1000, replace=TRUE)`)
    
sample_data = np.random.choice(list("abcdefghijklmnopqrstuvwxyz"), 1000, replace=True)
print("Mode of sample letters:", mode(sample_data))

# Test with binary target vector `y`

print("Mode of y:", mode(y))


Mode of sample letters: b
Mode of y: 0


Fit a threshold model to `y` using the feature `Sepal.Length`. Write your own code to do this. What is the estimated value of the threshold parameter? Save the threshold value as `threshold`. Hint: use np.zeros and np.sum from Numpy. You will need to use a for loop using the range() function.  

In [79]:
# Extract relevant data
sepal_length = df_iris_binary["sepal length (cm)"].values  # Feature
y_values = y.values  # Target labels (0 or 1)
n = len(sepal_length)  # Number of samples

# Initialize matrix to store threshold values and corresponding error counts
num_errors_by_parameter = np.zeros((n, 2))

# Loop over all possible threshold values
for i in range(n):
    threshold = sepal_length[i]  # Set current threshold
    num_errors = np.sum((sepal_length > threshold) != y_values)  # Count classification errors
    num_errors_by_parameter[i] = [threshold, num_errors]  # Store values

# Sort by number of errors
num_errors_by_parameter = num_errors_by_parameter[num_errors_by_parameter[:, 1].argsort()]

# Get the threshold with the least number of errors
best_threshold = num_errors_by_parameter[0, 0]

# Print results
print(f"Optimal threshold for classification: {best_threshold}")

Optimal threshold for classification: 5.4


What is the total number of errors this model makes? This requires a couple of minor modifications to the previous code.

In [83]:
# Extract relevant data
sepal_length = df_iris_binary["sepal length (cm)"].values  # Feature
y_values = y.values  # Target labels (0 or 1)
n = len(sepal_length)  # Number of samples

# Initialize matrix to store threshold values and corresponding error counts
num_errors_by_parameter = np.zeros((n, 2))
total_errors = 0
# Loop over all possible threshold values
for i in range(n):
    threshold = sepal_length[i]  # Set current threshold
    
    num_errors = np.sum((sepal_length > threshold) != y_values)  # Count classification errors
    
    # Store threshold and corresponding errors
    num_errors_by_parameter[i] = [threshold, num_errors]
    
    # Accumulate total errors across all thresholds
    total_errors += num_errors


# Sort by number of errors to find the best threshold
num_errors_by_parameter = num_errors_by_parameter[num_errors_by_parameter[:, 1].argsort()]
best_threshold = num_errors_by_parameter[0, 0]  # Best threshold with the least errors

# Print results
print(f"Optimal threshold for classification: {best_threshold}")
print(f"Total number of errors across all thresholds: {total_errors}")

Optimal threshold for classification: 5.4
Total number of errors across all thresholds: 2796


Does the threshold model's performance make sense given the following summaries:

In [84]:
# Print the best threshold found earlier
print(f"Optimal threshold for classification: {best_threshold}")

# Summary statistics for setosa and versicolor Sepal.Length
setosa_summary = df_iris_binary[df_iris_binary["species"] == 0]["sepal length (cm)"].describe()
versicolor_summary = df_iris_binary[df_iris_binary["species"] == 1]["sepal length (cm)"].describe()

# Print summaries
print("\nSummary statistics for Setosa Sepal Length:")
print(setosa_summary)

print("\nSummary statistics for Versicolor Sepal Length:")
print(versicolor_summary)

Optimal threshold for classification: 5.4

Summary statistics for Setosa Sepal Length:
count    50.00000
mean      5.00600
std       0.35249
min       4.30000
25%       4.80000
50%       5.00000
75%       5.20000
max       5.80000
Name: sepal length (cm), dtype: float64

Summary statistics for Versicolor Sepal Length:
count    50.000000
mean      5.936000
std       0.516171
min       4.900000
25%       5.600000
50%       5.900000
75%       6.300000
max       7.000000
Name: sepal length (cm), dtype: float64


TO-DO: Write your answer here in English

well setosa" ranges from 4.3 to 5.8, with a median of 5.0 and a mean of 5.006 and "versicolor" ranges from 4.9 to 7.0, with a median of 5.9 and a mean of 5.936 and threshold is 5.35 which makes sense because most of setosa is below it and most of versicolor is above it

Create the function `g` explicitly that can predict `y` from `x` being a new `Sepal.Length`. Hint: use np.where from Numpy ... this can also be down using a lambda function. 

In [91]:
# Define function `g` for threshold-based prediction

def g(x):
    return np.where(x > best_threshold, 1, 0)

#create test value 0 is less than threshold 1 is more than threshold
test_val = 12
g(test_val)

array(1)

In [93]:
#another way to do in one line

g = lambda x:np.where(x >best_threshold, 1, 0)
#create test value 0 is less than threshold 1 is more than threshold
test_val = 1
g(test_val)

array(0)